src/SpectatorView.Native/SpectatorView.Compositor/Compositor/AzureKinectCameraInput.cpp

```diff
@@ -0,0 +1,401 @@
+            }
+
+            // Create new depth texture for body depth only
+            k4a_image_create(K4A_IMAGE_FORMAT_DEPTH16, _calibration.depth_camera_calibration.resolution_width, _calibration.depth_camera_calibration.resolution_height, 2 * _calibration.depth_camera_calibration.resolution_width, &_bodyMaskImage);
+            k4a_image_create(K4A_IMAGE_FORMAT_DEPTH16, _calibration.color_camera_calibration.resolution_width, _calibration.color_camera_calibration.resolution_height, 2 * _calibration.color_camera_calibration.resolution_width, &_transformedBodyMaskImage);
+            _bodyMaskImageStride = k4a_image_get_stride_bytes(_transformedBodyMaskImage);
+        }
+#endif
+
+    }
+
+    _thread = std::make_shared<std::thread>(std::bind(&AzureKinectCameraInput::RunCaptureLoop, this));
+
+#if defined(INCLUDE_AZUREKINECT_BODYTRACKING)
+    if (captureBodyMask)
+    {
+        _bodyIndexThread = std::make_shared<std::thread>(std::bind(&AzureKinectCameraInput::RunBodyIndexLoop, this));
+    }
+#endif
+
+    return;
+
+FailedExit:
+    if (_k4aDevice != NULL)
+    {
+        k4a_device_close(_k4aDevice);
+    }
+}
+
+AzureKinectCameraInput::~AzureKinectCameraInput()
+{
+    _stopRequested = true;
+    _thread->join();
+
+#if defined(INCLUDE_AZUREKINECT_BODYTRACKING)
+    if (_bodyIndexThread != nullptr)
+    {
+        _bodyIndexThread->join();
+    }
+#endif
+
+    if (_bodyMaskImage != nullptr)
+    {
+        k4a_image_release(_bodyMaskImage);
+        _bodyMaskImage = nullptr;
+    }
+
+    if (_transformedBodyMaskImage != nullptr)
+    {
+        k4a_image_release(_transformedBodyMaskImage);
+        _transformedBodyMaskImage = nullptr;
+    }
+
+    if (_transformedDepthImage != nullptr)
+    {
+        k4a_image_release(_transformedDepthImage);
+        _transformedDepthImage = nullptr;
+    }
+
+    if (_k4aDevice != nullptr)
+    {
+        k4a_device_stop_cameras(_k4aDevice);
+        k4a_device_close(_k4aDevice);
+        _k4aDevice = nullptr;
+    }
+
+#if defined(INCLUDE_AZUREKINECT_BODYTRACKING)
+    if (_k4abtTracker != nullptr)
+    {
+        k4abt_tracker_shutdown(_k4abtTracker);
+        k4abt_tracker_destroy(_k4abtTracker);
+    }
+#endif
+
+    for (int i = 0; i < MAX_NUM_CACHED_BUFFERS; i++)
+    {
+        delete _cameraFrames[i];
+    }
+}
+
+void AzureKinectCameraInput::RunCaptureLoop()
+{
+    while (!_stopRequested)
+    {
+        int frameIndex = _currentFrameIndex % MAX_NUM_CACHED_BUFFERS;
+        if (!_cameraFrames[frameIndex]->TryBeginWritingColorAndDepth())
+        {
+            // If the next frame in the buffer is still pending write or is being read from, then we've
+            // exceeded the capacity of the buffer and either the reader is blocking too long, or
+            // the body index processing thread has fallen behind.
+            OutputDebugString(L"Warning: frame buffer is completely full, and we can't begin writing to the next frame");
+            continue;
```

Comment:
do we think that the color/depth capture should have authority here? Or is having the body tracking more important than keeping the color frame up to date?

Reply:
another solution here seems to be to clear the body tracking data here compared to dropping the color/depth frame. I'm not sure which is worse.

---
In reply to: [403197138](https://github.com/microsoft/MixedReality-SpectatorView/pull/356#discussion_r403197138) [](ancestors = 403197138)

Reply:
I think that leaving the current behavior for now is ok, until we hit a case where the body tracking fails to keep up with the RGB camera. The result currently will be that the camera might fall below 30 frames per second, at which point the recorded video might have slight time slowdown.

---
In reply to: [403198425](https://github.com/microsoft/MixedReality-SpectatorView/pull/356#discussion_r403198425) [](ancestors = 403198425,403197138)